### PR TITLE
[V5] Cache loader improvements

### PR DESCRIPTION
**Pull request by erikn69**
- Deleted old V4 fallbacks
- Early return used
- Avoid use same bucle twice on permisions (`collect($this->permissions)->map()` and later `->each()`)
- `array_map` instead of `collect($this->permissions)->map()->all()` and other `collect()` again with the first collect

### CODE REVIEW

**Comment by freekmurze** [comment]
To me, the old and new code is unreadable. Could you split this up in dedicated functions that are clearly named and explain what is happening?

**Comment by erikn69** [comment]
@freekmurze Fixed

**Comment by erikn69** [comment]
I have another commit that makes cache 50% smaller by serialization of Role model ids, 
but I have not uploaded it because I know you like the changes in parts, so I waiting to see if it is ok

TamanoCache

**Comment by freekmurze** [comment]
Will these cache improvements break if people have already cached things using the old way?

**Comment by erikn69** [comment]
No breaks, there is a `if` that forget cache if it is old versions, a temporal fix for that

File: src/PermissionRegistrar.php [hunk 1]
```diff
@@ -168,32 +168,22 @@ private function loadPermissions()
...
-            $this->cachedRoles = [];
+        // fallback for old cache method, must be removed on next mayor version
+        if (! isset($this->permissions['permissions'])) {
+            $this->forgetCachedPermissions();
+            $this->loadPermissions();
+
+            return;
         }
+
+        $this->hidrateRolesCache();
```
Comment (Restingo) [inline src/PermissionRegistrar.php:182]:
```suggestion
        $this->hydrateRolesCache();
```

File: src/PermissionRegistrar.php [hunk 2]
```diff
@@ -275,21 +265,82 @@ public function getCacheStore(): \Illuminate\Contracts\Cache\Store
...
+    }
+
+    private function getHydratedRoleCollection(array $roles)
+    {
+        return Collection::make(array_values(
+            array_intersect_key($this->cachedRoles, array_flip($roles))
+        ));
+    }
+
+    private function hidrateRolesCache()
```
Comment (Restingo) [inline src/PermissionRegistrar.php:333]:
```suggestion
    private function hydrateRolesCache()
```

**Review by Restingo — COMMENTED** [review]
Typo

**Comment by angeljqv** [comment]
@freekmurze will this PR be merged?

**Comment by PaolaRuby** [comment]
.

**Comment by angeljqv** [comment]
@freekmurze will this PR be merged?

**Comment by freekmurze** [comment]
Thanks! 👍

**Comment by erikn69** [comment]
Thank you, i am getting better times and less ram usage
image

**Pull request merged**